Updates to brute force method

reneeotten opened the pull request:
This PR includes two updates to the brute force method:

1) fix the issue that pickle doesn't work with the `MinimizerResult` when the brute force method is used (related to the use of a `namedtuple`).

2) add a function to visualize the result of the grid search (inspired by the way the `corner` package can be used to visualize the results of the `emcee` method). Comments and suggestions on the actual code are welcome, as well as where to put it; right now I added it into a file called `plotfuncs.py`.

Before a possible merge, it will need some updates to the documentation. Example has been updated together with the resulting output file (`examples/results_brute.pdf`), and no tests were added or changed.

newville commented:
@reneeotten   Thanks!  This seems like 2 slightly different topics to me.   On "fix pickle with namedtuple", I think there's no question that this should be done.

For plotting utilities, I am all in favor of better data visualization, but I worry about adding plotting capabilities to lmfit as it could easily dominate the project.  I am in favor of a `plotfunc.py` module for adding such nice visualizations of fits and so on, but we might want to err on the side of "simple / obvious" rather than try to accommodate all possible options. 

I think the plotting capabilities in `ModelResult`were very nicely done, and should probably be the model (ouch!) to follow here.  Some of those nice features to consider emulating (or discuss):

  1.  checking for the import of matplotlib, using the `_ensureMatplotlib` decorator.
  2.  not actually producing output plots or files but returning `Axes` or other matplotlib things that can be modified and/or plotted by the user. 

We could think about moving some of the functionality of `ModelResult.plot_*` to `plotfuncs.py` and then have those methods call the more general functions, but perhaps not now.

It might be helpful to come up with a set of functions (like those in `ModelResult`) that are the most common ways to visualize fit results.  "Data+Fit with Residual sharing an X axis", "maps of chi-square (or cost) for correlated parameters", and "corner plots" seem likely candidates to me.  
I would not say that any of those are needed for this PR, but it's worth thinking about what would be desirable.

Any other opinions on this?

reneeotten commented:
@newville thanks for the comments. I totally agree with you on the fact that `lmfit` should be about fitting and not dominated by plotting functionalities. I think that defining some minimal plotting functionalities could be useful and would warrant some further thought on how to implement that. Right now I don't think I'll have time for that though...

Why don't I change this PR such that it fixes the `pickle` issue and move the plotting functions into the example file. I'll use (and possible refine) it a bit more in my own use of it and then we can decide later on how/where to add to the main code. Okay?

newville commented:
@reneeotten Yes,  I agree taking out the plotting code and expecting that to be part of a (slightly) larger conversation / code would be better.   To be clear, I'm completely OK with having some plotting functions, especially for non-trivial or not-quite-simple things like corner plots. 

Also, with this PR and a couple other recent / pending PRs we have some important bug fixes to 0.9.6.   Anyone have any thoughts on or objections to releasing these as 0.9.7 soon?

reneeotten commented:
alright - I have updated the PR. It fixes the pickle vs namedtuple issue and includes a test for it. I have moved the visualization/plotting code to the `examples/lmfit_brute.py` file for now.

newville commented:
@reneeotten Great - thanks!  Anyone have any objections to this?
